cpp/src/io/parquet/experimental/dictionary_page_filter.cu

```diff
@@ -46,11 +48,1527 @@ using parquet::detail::ColumnChunkDesc;
+      // Check if we are overruning the data stream
+      if (is_stream_overrun(value_idx * sizeof(T), sizeof(T), page.uncompressed_page_size)) {
+        set_error(error, decode_error::DATA_STREAM_OVERRUN);
+        return {};
+      }
+      cuda::std::memcpy(&decoded_value, page_data + (value_idx * sizeof(T)), sizeof(T));
+      break;
+    }
+    default: {
+      // Parquet physical type is not fixed width so set the error code and break early
+      set_error(error, decode_error::INVALID_DATA_TYPE);
+      return {};
+    }
+  }
+
+  return decoded_value;
+}
+
+/**
+ * @brief Decode a string value from a page data buffer
+ *
+ * @param page_data Pointer to the page data buffer
+ * @param page_data_size Page data buffer size
+ * @param buffer_offset Current offset into the page data buffer
+ * @param error Pointer to the kernel error code
+ * @return Decoded string value
+ */
+__device__ cudf::string_view decode_string_value(uint8_t const* page_data,
+                                                 int32_t page_data_size,
+                                                 int32_t& buffer_offset,
+                                                 kernel_error::pointer error)
+{
+  if (is_stream_overrun(buffer_offset, sizeof(int32_t), page_data_size)) {
+    set_error(error, decode_error::DATA_STREAM_OVERRUN);
+    return {};
+  }
+
+  // Decode string length
+  auto const string_length = static_cast<int32_t>(*(page_data + buffer_offset));
+  buffer_offset += sizeof(int32_t);
+
+  // Check if we have a stream overrun
+  if (is_stream_overrun(buffer_offset, string_length, page_data_size)) {
+    set_error(error, decode_error::DATA_STREAM_OVERRUN);
+    return {};
+  }
+
+  // Decode cudf::string_view value
+  auto const decoded_value =
+    cudf::string_view{reinterpret_cast<char const*>(page_data + buffer_offset),
+                      static_cast<cudf::size_type>(string_length)};
+
+  // Update the buffer offset
+  buffer_offset += string_length;
+
+  return decoded_value;
+}
+
+/**
+ * @brief Decode string column chunk dictionaries and build `cuco::static_set`s from them,
+ * one hash set per dictionary
+ *
+ * @param pages Column chunk dictionary page headers
+ * @param decoded_data Span of storage for decoded values from all dictionaries
+ * @param set_storage Pointer to the start of the bulk cuco hash set slots
+ * @param set_offsets Pointer to offsets into the bulk set storage for each dictionary
+ * @param value_offsets Pointer to offsets into running sum of values in each dictionary
+ * @param num_dictionary_columns Total number of columns with dictionaries
+ * @param dictionary_col_idx Index of the current dictionary column
+ * @param error Pointer to the kernel error code
+ */
+CUDF_KERNEL void __launch_bounds__(DECODE_BLOCK_SIZE)
+  build_string_dictionaries(PageInfo const* pages,
+                            cudf::device_span<cudf::string_view> decoded_data,
+                            slot_type* const set_storage,
```

Comment:
Hate to nitpick, but why this only this parameter const?

Reply:
We can modify contents at `slot_type*` but just marked the ptr itself to be const. Could just mark most of them as const too

Reply:
Alrighty, made it non-const and reverted the ones previously marked const